c_src/libSigPyProc.cpp

```diff
@@ -0,0 +1,674 @@
+            int   pos  = nchans * ii * tfactor + jj * ffactor;
+            float temp = 0;
+            for (int kk = 0; kk < tfactor; kk++) {
+                for (int ll = 0; ll < ffactor; ll++) {
+                    temp += indata[kk * nchans + ll + pos];
+                }
+            }
+            float result = temp / totfactor;
+            if (std::is_same<T, uint8_t*>::value) {
+                result = result + 0.5;
+            }
+            outdata[ii * newnchans + jj] = (T)(result);
+        }
+    }
+}
+
+/**
+ * Remove the channel-weighted zero-DM (Eatough, Keane & Lyne 2009)
+ * code based on zerodm.c (presto)
+ */
+template <class T>
+void removeZeroDM(py::array_t<T> inarray, py::array_t<T> outarray,
+                  py::array_t<float> bpass, py::array_t<float> chanwts,
+                  int nchans, int nsamps) {
+    py::buffer_info inbuf = inarray.request(), outbuf = outarray.request(),
+                    bpassbuf = bpass.request(), chanwtsbuf = chanwts.request();
+
+    T*     indata      = (T*)inbuf.ptr;
+    T*     outdata     = (T*)outbuf.ptr;
+    float* bpass_arr   = (float*)bpassbuf.ptr;
+    float* chanwts_arr = (float*)chanwtsbuf.ptr;
+
+#pragma omp parallel for default(shared)
+    for (int ii = 0; ii < nsamps; ii++) {
+        double zerodm = 0.0;
+        for (int jj = 0; jj < nchans; jj++) {
+            zerodm += indata[(nchans * ii) + jj];
+        }
+        for (int jj = 0; jj < nchans; jj++) {
+            if (std::is_same<T, uint8_t*>::value) {
+                zerodm = zerodm + 0.5;
+            }
+            outdata[(nchans * ii) + jj] =
+                (T)((indata[(nchans * ii) + jj] - zerodm * chanwts_arr[jj]) +
+                    bpass_arr[jj]);
+        }
+    }
+}
+
+
+
+PYBIND11_MODULE(libSigPyProc, m) {
+    m.doc() = "libSigPyProc functions";
+
+    m.def("_omp_get_max_threads", 
+        []() { return omp_get_max_threads(); });
+    m.def("_omp_set_num_threads", 
+        [](int nthread) { omp_set_num_threads(nthread); });
+
+    m.def("unpack", &unpack, 
+        "Unpack 1, 2 and 4 bit data into an 8-bit numpy array",
+        py::arg("inarray"), py::arg("nbits"));
+
+    m.def("unpackInPlace", &unpackInPlace, 
+        "Unpack 1, 2 and 4 bit data into 8-bit in the same numpy array",
+        py::arg("inarray"), py::arg("nbits"));
+    
+    m.def("pack", &pack, 
+        "Pack 1, 2 and 4 bit data into an 8-bit numpy array",
+        py::arg("inarray"), py::arg("nbits"));
+
+    m.def("packInPlace", &packInPlace, 
+        "Pack 1, 2 and 4 bit data into 8-bit in the same numpy array",
+        py::arg("inarray"), py::arg("nbits"));
+
+    m.def("to8bit", &to8bit, 
+        "Convert 1, 2 and 4 bit data to 8-bit",
+        py::arg("inarray"), py::arg("outarray"), py::arg("flag"), 
+        py::arg("fact"), py::arg("plus"), py::arg("flagMax"), 
+        py::arg("flagMin"), py::arg("nsamps"), py::arg("nchans"));
+
+    m.def("splitToBands", &splitToBands<float>);
```

Comment:
Note: I have never used this with templates -- I'm assuming this does what you want and exposes 8-bit and floating point? Wondering whether float32 and float64 should be made explicit, but think the effect would be the same so may be pointless.

Reply:
Yes, for templates, we need to instantiate it for all types explicitly. 
No float64 in python API yet, so no need.